### PR TITLE
Add Terraform plan processing support

### LINKED ISSUE #1193: Support for Terraform
This is tracking for adding support for Terraform HCL to PSRule originally discussed here Microsoft/PSRule#907 by @jbla9028.
---

**Is your feature request related to a problem? Please describe.**
It appears that psrule can review a ARM or bicep template from the documentation. It seems like the only open to review terraform is to review the resources after the terraform code has applied the infrastructure. Can we get native terraform support?

Passing a target folder, along with variables, can psrule review a user's terraform code for compliance?

**Describe the solution you'd like**

**Describe alternatives you've considered**
There are other solutions that will do this but psRule's syntax seems very easy to use, it would be great to leverage it with terraform.

**Additional context**

### PR DESCRIPTION
**Pull request by polatengin**
## PR Summary

- Introduced TerraformPlanException for handling errors related to Terraform plan files.
- Implemented TerraformPlanHelper to process Terraform plan JSON files and extract Azure resources.
- Enhanced the Helper class to check for Terraform plan files and expand resources from them.
- Updated configuration to enable Terraform plan expansion.
- Added conventions for expanding Terraform plans in the Conventions.Rule.ps1.
- Included test cases for Terraform plan processing, covering various scenarios and edge cases.
- Added sample Terraform plan JSON files for testing purposes.

Fixes #1193 

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [x] Unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: docs/using-terraform.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+Analyze the plan JSON file with PSRule:
+
+```bash
+Assert-PSRule -InputPath './tfplan.json' -Module PSRule.Rules.Azure
+```
+
+Or using the CLI:
+
+```bash
+ps-rule run -InputPath './tfplan.json' --module PSRule.Rules.Azure
```
Comment (BernieWhite) [inline docs/using-terraform.md]:
```suggestion
ps-rule run --input-path './tfplan.json' --module PSRule.Rules.Azure
```

File: docs/using-terraform.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+configuration:
+  AZURE_TERRAFORM_PLAN_EXPANSION: true
+```
+
+### Step 3: Run PSRule
+
+Analyze the plan JSON file with PSRule:
+
+```bash
+Assert-PSRule -InputPath './tfplan.json' -Module PSRule.Rules.Azure
```
Comment (BernieWhite) [inline docs/using-terraform.md]: `-Format File` is required for this command line.

```suggestion
Assert-PSRule -InputPath './tfplan.json' -Format File -Module PSRule.Rules.Azure
```

File: docs/using-terraform.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+- **`azapi_data_plane_resource`** &mdash; Data plane operations don't map to ARM resources.
+- **Non-Azure providers** &mdash; Resources from `hashicorp/random`, `hashicorp/null`, etc. are skipped.
+
+## Limitations
+
+The following limitations apply to Terraform plan expansion:
+
+- Only `azapi_resource` and `azapi_update_resource` types are supported. The `azurerm` provider is not yet supported.
+- Raw `.tf` files cannot be analyzed directly. You must first generate a plan JSON file.
+- **Unknown-at-plan-time values** &mdash; Properties that Terraform cannot resolve until after `apply` (e.g., resource IDs, FQDNs) will be absent from analysis.
```
Comment (BernieWhite) [inline docs/using-terraform.md:104]: Noticed this with using an expression from a Azure Learn page. I.e. `coalesce(var.container_registry_name, random_string.acr_name.result)` won't work if it falls through to the random string.

```suggestion
- **Unknown-at-plan-time values** &mdash; Properties that Terraform cannot resolve until after `apply` (e.g., resource IDs, FQDNs) will be absent from analysis.
  - As a result, using the `hashicorp/random` to generate resource names dynamically is currently not supported.
```

File: docs/using-terraform.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+- **Data sources** (`mode: "data"`) &mdash; Read-only lookups, not deployable resources.
+- **`azapi_resource_action`** &mdash; Invokes POST actions, not a resource definition.
+- **`azapi_data_plane_resource`** &mdash; Data plane operations don't map to ARM resources.
+- **Non-Azure providers** &mdash; Resources from `hashicorp/random`, `hashicorp/null`, etc. are skipped.
+
+## Limitations
+
+The following limitations apply to Terraform plan expansion:
+
+- Only `azapi_resource` and `azapi_update_resource` types are supported. The `azurerm` provider is not yet supported.
```
Comment (BernieWhite) [inline docs/using-terraform.md:102]: Not a problem for this PR, but a follow up. I noticed a lot of Learn docs including AVM reference `azurerm_resource_group`.

While I don't think we need to tackle all azurerm resource as this stage, I think we should add support for this one as early as feasible. An since it has minimal properties it should be much easier.

File: tests/PSRule.Rules.Azure.Tests/TerraformPlanHelperTests.cs [hunk 5]
```diff
@@ -0,0 +1,461 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/TerraformPlanHelperTests.cs]: nit: Not needed.

```suggestion
```

File: docs/changelog.md [hunk 6]
```diff
@@ -32,6 +32,11 @@ See [upgrade notes][1] for helpful information when upgrading from previous vers
 
 What's changed since v1.47.0:
 
+- New features:
+  - Added Terraform plan expansion support for AzAPI provider resources by @polatengin.
```
Comment (BernieWhite) [inline docs/changelog.md]:
```suggestion
  - **Experimental:** Added Terraform plan expansion support for AzAPI provider resources by @polatengin.
```

**Review by BernieWhite — CHANGES_REQUESTED** [review]
Thanks @polatengin. Great work!

A few minor changes and notes below.

Also lets add details on `AZURE_TERRAFORM_PLAN_EXPANSION` to `docs\setup\configuring-expansion.md`

Otherwise we're good to merge.

File: docs/using-terraform.md (hunk 4 above)
Comment (polatengin) [reply docs/using-terraform.md:102]: I'll create a follow up PR with basic azurerm support, after this one merged 👍

**Review by BernieWhite — DISMISSED** [review]
Thanks @polatengin, great work. There is a conflict on change log. Otherwise all good to merge.

**Review by BernieWhite — APPROVED** [review]
